takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java

```diff
@@ -0,0 +1,81 @@
+package io.takari.maven.plugins;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.maven.AbstractMavenLifecycleParticipant;
+import org.apache.maven.MavenExecutionException;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.plugin.descriptor.PluginDescriptor;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.logging.LogEnabled;
+import org.codehaus.plexus.logging.Logger;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.deployment.DeployRequest;
+import org.eclipse.aether.deployment.DeploymentException;
+
+public class DeployParticipant extends AbstractMavenLifecycleParticipant implements LogEnabled {
+
+  private static final String CONTEXT_KEY = "deploy-at-end-requests";
+
+  protected RepositorySystem repoSystem;
+  private Logger logger;
+
+  @Override
+  public void enableLogging(Logger logger) {
+    this.logger = logger;
+  }
+
+  @Override
+  public void afterSessionEnd(MavenSession session) throws MavenExecutionException {
+    List<DeployRequest> deployAtEndRequests = getDeployAtEndRequests(session);
+    boolean noerrors = session.getResult().getExceptions().isEmpty();
+
+    if (!deployAtEndRequests.isEmpty() && noerrors) {
+      logger.info("");
+      logger.info("------------------------------------------------------------------------");
+      logger.info("-- Performing deploy at end                                           --");
+      logger.info("------------------------------------------------------------------------");
+
+      synchronized (deployAtEndRequests) {
+        for (DeployRequest deployRequest : deployAtEndRequests) {
+          try {
+            repoSystem.deploy(session.getRepositorySession(), deployRequest);
+          } catch (DeploymentException e) {
+            logger.error(e.getMessage(), e);
+            throw new MavenExecutionException(e.getMessage(), e);
+          }
+        }
+        deployAtEndRequests.clear();
+      }
+
+      logger.info("------------------------------------------------------------------------");
+    }
+  }
+
+  public static void deployAtEnd(MavenSession session, DeployRequest deployRequest) {
+    getDeployAtEndRequests(session).add(deployRequest);
+  }
+
+  public static List<DeployRequest> getDeployAtEndRequests(MavenSession session) {
+    MavenProject topLevelProject = session.getTopLevelProject();
```

Comment:
I think it is better to keep deployment request list in an instance variable. It will make the code simpler to understand and maintain. It will also work better in the unlikely event when the same reactor has multiple versions of takari-lifecycle.